### PR TITLE
[Frontend] don't block event loop in tokenization (preprocess) in OpenAI compatible server

### PR DESCRIPTION
**Pull request by tomeras91**
Currently, tokenization of requests in the OpenAI compatible server is done synchronously on the process running the server. Usually, this is not a problem, but when dealing with long sequence lengths (256K), tokenization can take a few hundred milliseconds. Combining long requests with high loads causes the server to be non-responsive as it spends a lot of time tokenizing requests synchronously and sequentially, blocking the asyncio event loop.

This PR aims to unblock the event loop by using a thread pool to tokenize requests. This following https://github.com/vllm-project/vllm/pull/3512 which also introduced parallel async tokenization but was closed and this feature was never delivered.

### CODE REVIEW

File: vllm/commit_id.py [hunk 1]
```diff
@@ -0,0 +1 @@
+__commit__ = "534e75c1193165605aee6d741deb4a1536c58ff1"
```
Comment (robertgshaw2-neuralmagic) [inline vllm/commit_id.py:1]: nit

**Comment by robertgshaw2-neuralmagic** [comment]
LGTM - as a sanity check, can you run a quick ShareGPT benchmark on a small model?

File: vllm/entrypoints/openai/serving_engine.py [hunk 2]
```diff
@@ -397,7 +397,52 @@ def _tokenize_prompt_input_or_inputs(
...
+        input_or_inputs: Union[str, List[str], List[int], List[List[int]]],
+        truncate_prompt_tokens: Optional[Annotated[int, Field(ge=1)]] = None,
+        add_special_tokens: bool = True,
+    ) -> Iterator[TextTokensPrompt]:
+        return await make_async(self._tokenize_prompt_input_or_inputs)(
+            request=request,
+            tokenizer=tokenizer,
+            input_or_inputs=input_or_inputs,
+            truncate_prompt_tokens=truncate_prompt_tokens,
+            add_special_tokens=add_special_tokens)
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/serving_engine.py:443]: I think it would be simpler if we just call `make_async` and assign these methods in `__init__`.

Comment (tomeras91) [reply vllm/entrypoints/openai/serving_engine.py:443]: got it. Like in [multiproc_gpu_executor.py](https://github.com/vllm-project/vllm/blob/ed46f143212203b7afcbc8538119b6e8155c643e/vllm/executor/multiproc_gpu_executor.py#L219) for example.

**Comment by tomeras91** [comment]
> LGTM - as a sanity check, can you run a quick ShareGPT benchmark on a small model? If you need instructions for this, I can share commands.

Sure. Here are the results

Model: Qwen/Qwen2.5-1.5B-Instruct
Hardware: single H100
| branch | throughput (requests/s) | mean TTFT (msec) |
|--------|--------|--------|
| main (commit ed46f143212203b7afcbc8538119b6e8155c643e) | 51.33 | 6645.88 |
| pr (commit  980fff886f93e300a456b2ba05e6ffe95e52c0d2) | 51.62 | 6251.77 |

So results are pretty much identical. Maybe a bit better with the PR but I guess the diffs are too small to be meaningful.

Other than that, I think the CI failures are not related to the changes in this PR

**Review by njhill — COMMENTED** [review]
Thanks @tomeras91.

Another thought is that it's likely optimal to only dispatch to another thread if the length of text / number of tokens is above some threshold, otherwise it may be slightly detrimental.

But like @robertgshaw2-neuralmagic said, we should make sure to benchmark this anyhow.

**Comment by njhill** [comment]
@tomeras91 presumably you'll also test with your >200k token workload to ensure that it helps?

File: vllm/entrypoints/openai/serving_engine.py [hunk 3]
```diff
@@ -140,6 +140,13 @@ def __init__(
         self.request_logger = request_logger
         self.return_tokens_as_token_ids = return_tokens_as_token_ids
 
+        self._tokenize_prompt_input_async = make_async(
+            self._tokenize_prompt_input)
+        self._tokenize_prompt_inputs_async = make_async(
+            self._tokenize_prompt_inputs)
+        self._tokenize_prompt_input_or_inputs_async = make_async(
+            self._tokenize_prompt_input_or_inputs)
```
[the remark this answers is not in the document]
Comment (tomeras91) [reply vllm/entrypoints/openai/serving_engine.py]: that's a good catch. It really didn't work for /v1/completions which uses these methods. It did work for /v1/chat/completions that uses `_tokenize_prompt_input` which doesn't return a generator and actually does the tokenization work.
Anyway, fixed by making `_tokenize_prompt_input_or_inputs` return a `List` as you suggested.

Will post updated benchmarks and something that shows this works shortly

**Comment by tomeras91** [comment]
### Updated benchmarks
Now both on `/v1/completions` and on `/v1/chat/completions`

Model: Qwen/Qwen2.5-1.5B-Instruct
Hardware: single H100
Serve run command: `vllm serve Qwen/Qwen2.5-1.5B-Instruct`

Endpoint: `v1/completions`
(benchmark run command: `python3 benchmark_serving.py --model $MODEL --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json`

| branch | throughput (requests/s) | mean TTFT (msec) | mean TPOT (msec) | Total generated tokens |
|--------|--------|--------|--------|--------|
| main (commit ed46f143212203b7afcbc8538119b6e8155c643e) | 52.88 | 6406.30 | 23.09 | 189265 |
| pr (commit  b61a04ff01c5f5990e0335de16d214620f09169e) | 52.92 | 5985.69 | 21.29 | 191136 |

Conclusion: Pretty much the same (very slightly better performance in PR even though more tokens were generated)

Endpoint: `v1/chat/completions`
(benchmark run command: `python3 benchmark_serving.py --model $MODEL --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json  --endpoint /v1/chat/completions --backend openai-chat`
| branch | throughput (requests/s) | mean TTFT (msec) | mean TPOT (msec) | Total generated tokens |
|--------|--------|--------|--------|--------|
| main (commit ed46f143212203b7afcbc8538119b6e8155c643e) | 57.63 | 5946.73 | 23.73 | 171968 |
| pr (commit  b61a04ff01c5f5990e0335de16d214620f09169e) | 55.52 | 5972.38 | 23.50 | 172531 |

Conclusion: Pretty much the same (slightly worse performance in PR, maybe due to more generated tokens)

Anyway, I would expect this PR to affect TTFT much more than it affects TPOT, but it seems like a very insignificant change regardless. If at all, the effect is more significant for the completions endpoint, and is an improvement.


### RE something to show this actually solves the problem:
Here is the code I used:
```python
import requests
import threading
from threading import Thread
import time

def send_request(url, body):
    # print(f"Sending from thread {threading.get_ident()}")
    start = time.time()
    res = requests.post(url, json=body)
    end = time.time()
    # print(f"got res in thread {threading.get_ident()} after {1000*(end-start):.2f} ms with status {res.status_code}. content: {res.json() if res.status_code!=200 else 'success'}")
    return res

model = "Qwen/Qwen2.5-1.5B-Instruct"

chat_url = "http://localhost:8000/v1/chat/completions"
chat_body = {'model': model,
             'messages': [{'role': 'user', 'content': ' '.join(['A']*300_000)}],
            }

completions_url = "http://localhost:8000/v1/completions"
completions_body = {'model': model,
                    'prompt': ' '.join(['A']*300_000),
                   }
completions_body_mult = {'model': model,
                         'prompt': [' '.join(['A']*300_000)]*2,
                        }

for _ in range(60):
    # Thread(target=send_request, kwargs={'url': completions_url, 'body': completions_body}).start()
    Thread(target=send_request, kwargs={'url': completions_url, 'body': completions_body_mult}).start()
    # Thread(target=send_request, kwargs={'url': chat_url, 'body': chat_body}).start()
start = time.time()
res = requests.get("http://localhost:8000/health")
end = time.time()
print(f"######################### healthcheck took {end-start} secs #########################")
```
The idea is to send many long context requests in close succession (using multiple threads), and then check how long `/health` takes to respond. If the event loop is blocked, we would expect long response time of the healthcheck. As you can see, I checked it for both `v1/completions` and `v1/chat/completions`, and also for multiple inputs for the completions endpoint. Note that these requests are longer than the supported max model length (32K) so no work is actually done by the model and the requests are blocked after tokenization.

The response time of `/health` when the server is not dealing with any requests was about 2.3 msecs
| branch | endpoint | health response time (msec) |
|--------|--------|--------|
| main (commit ed46f143212203b7afcbc8538119b6e8155c643e) | `v1/completions` |  17848.1 |
| main (commit ed46f143212203b7afcbc8538119b6e8155c643e) | `v1/completions`  multiple inputs | 14285.4 |
| main (commit ed46f143212203b7afcbc8538119b6e8155c643e) | `v1/chat/completions`  multiple inputs | 17159.3 |
| pr (commit  b61a04ff01c5f5990e0335de16d214620f09169e) | `v1/completions` | 34.6 |
| pr (commit  b61a04ff01c5f5990e0335de16d214620f09169e) | `v1/completions`  multiple inputs | 41.0 |
| pr (commit  b61a04ff01c5f5990e0335de16d214620f09169e) | `v1/chat/completions`  multiple inputs | 16.4 |

The response time is much shorter in the PR. We do see that the healthcheck takes a bit longer to respond under this long context requests load, which can be attributed to the fact that tokenizing many such long requests does take a bit CPU

**Comment by njhill** [comment]
Thanks @tomeras91, it looks great now. One minor concern I realized is that the HF tokenizers technically aren't thread-safe. I think in practice this is only the case if things like truncation and/or padding are being done, but that can be true for us (if truncate parameter is specified in the request).

Despite this it might still not be a problem given that I don't think the GIL is really released. But I'm not sure whether we should put an explicit lock around the calls to the tokenizer or create a tokenizer per thread. The former would likely be less invasive and could be done by wrapping/replacing the `__call__` method when the tokenizer is created. It would probably be best to re-run the perf test after doing that.

**Comment by njhill** [comment]
Thanks @tomeras91 this looks great. Unfortunately it looks like there's one more problem (seen via failing CI test)...

The tokenizer gets included in logitsprocessors used for guided decoding. We currently initialize these in the front-end process and they are pickled and sent to the back-end .. which is a bit crazy but needed to avoid their creation blocking the GIL in the main proc (unpickling them is faster). There are plans to change this but in the meantime this change breaks things because the lock can't be picked.

Ideas to fix this:
1. Use (potentially separately constructed) non-threadsafe tokenizer for these logits processors
2. Implement this in your threadsafe tokenizer class https://docs.python.org/3/library/pickle.html#handling-stateful-objects 
3. Another idea I thought of which might be simpler overall - use a custom ThreadPoolExecutor with max_workers=1 for the make_async calls. This would avoid needing the lock provided that all the tokenize calls for the tokenizer are done async

I think I like (2) the least ... just makes this workaround more complex/convoluted.

**Comment by tomeras91** [comment]
Yeah I saw the failing CI but didn't get to fixing it yet.
I thought of fixing it by trying to extract the underlying non-threadsafe tokenizer before sending it to `outlines`. I saw there's already a "adapt tokenizer" function so I can add a bit more logic there.

I think I'll try your idea (3) as well. It makes a lot of sense as there's no real gain in running the tokenizer in multiple threads and it will remove the need for the lock. 

I also want to lock some of these behaviors (short healthcheck response time under heavy long context load, no failing with mix of truncated and not truncated requests) in unittests. Still don't know how hard will that be but given there are many tests that use an OpenAI server I think it shouldn't be too hard

**Comment by tomeras91** [comment]
yep. Making the tokenization async with a `ThreadPoolExecutor(max_workers=1)` works perfectly 🙂 
I'll update the PR and add new benchmarks and hopefully some unittests

File: tests/entrypoints/openai/test_async_tokenization.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        }),
+        (lambda x: x.chat.completions.create, {
+            "messages": [{
+                "role": "user",
+                "content": " ".join(['A'] * 10_000)
+            }]
+        }),
+    ],
+)
+async def test_with_and_without_truncate(
```
Comment (tomeras91) [inline tests/entrypoints/openai/test_async_tokenization.py:58]: This fails if we use a threadpool with max_workers>1 for tokenization because the tokenizer is not threadsafe. It passes both on `main` and on the PR

File: tests/entrypoints/openai/test_async_tokenization.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        }),
+        (lambda x: x.chat.completions.create, {
+            "messages": [{
+                "role": "user",
+                "content": " ".join(['A'] * 300_000)
+            }]
+        }),
+    ],
+)
+async def test_healthcheck_response_time(
```
Comment (tomeras91) [inline tests/entrypoints/openai/test_async_tokenization.py:108]: This fails on `main` and passes on the PR. It validates the fix actually works and the server event loop isn't blocked on tokenization

**Review by tomeras91 — COMMENTED** [review]
@njhill - Now we use a threadpool with max_workers=1. 

Also added tests. I put them in a new file because I couldn't find a better place where they belong. If you want me to move them elsewhere, that's not a problem of course

**Comment by tomeras91** [comment]
### Updated (hopefully last) Benchmarks
I ran the same sort of benchmarks as in https://github.com/vllm-project/vllm/pull/10635#issuecomment-2498885437

Model: Qwen/Qwen2.5-1.5B-Instruct
Hardware: single H100
Serve run command: `vllm serve Qwen/Qwen2.5-1.5B-Instruct`

Endpoint: `v1/completions`
(benchmark run command: `python3 benchmark_serving.py --model $MODEL --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json`

| branch | throughput (requests/s) | mean TTFT (msec) | mean TPOT (msec) | Total generated tokens |
|--------|--------|--------|--------|--------|
| main (commit ed46f143212203b7afcbc8538119b6e8155c643e) | 52.88 | 6406.30 | 23.09 | 189265 |
| pr (commit b61a04ff01c5f5990e0335de16d214620f09169e) | 52.92 | 5985.69 | 21.29 | 191136 |
| pr threadsafe (commit e59cc81500250c37987f205398c6945b03010988) | 52.75 | 6090.83 | 24.06 | 190028 |
| pr single thread threadpool (commit ff1d6a956e0d09b327480a984685f118a4f3ef87) | 52.66 | 6092.80 | 23.57 | 190356 |

Endpoint: `v1/chat/completions`
(benchmark run command: `python3 benchmark_serving.py --model $MODEL --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json  --endpoint /v1/chat/completions --backend openai-chat`
| branch | throughput (requests/s) | mean TTFT (msec) | mean TPOT (msec) | Total generated tokens |
|--------|--------|--------|--------|--------|
| main (commit ed46f143212203b7afcbc8538119b6e8155c643e) | 57.63 | 5946.73 | 23.73 | 171968 |
| pr (commit b61a04ff01c5f5990e0335de16d214620f09169e) | 55.52 | 5972.38 | 23.50 | 172531 |
| pr threadsafe (commit e59cc81500250c37987f205398c6945b03010988) | 55.83 | 5782.09 | 25.14 | 171961 |
| pr single thread threadpool (commit ff1d6a956e0d09b327480a984685f118a4f3ef87) | 55.68 | 5851.25 | 24.94 | 171941 |

**same conclusions** - there's no hit to performance

**`/health` response times** are <~5 ms. I also added a unittest to make sure they response time under high tokenization load are not too different than under no load.

**Review by njhill — APPROVED** [review]
Thanks @tomeras91!
